filters/args.go

```diff
@@ -0,0 +1,181 @@
+package filters
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+)
+
+type FilterArgs struct {
+	args []interface{}
+	pos  int
+	errs []string
+}
+
+func Args(args []interface{}) *FilterArgs {
+	return &FilterArgs{args: args}
+}
+
+func (a *FilterArgs) String() (_ string) {
+	if x, ok := a.next(); ok {
+		if s, ok := x.(string); ok {
+			return s
+		}
+		a.err("%v is not a string", x)
+	}
+	return
+}
+
+func (a *FilterArgs) StringOr(defValue string) string {
+	if a.pos >= len(a.args) {
+		return defValue
+	}
+	return a.String()
+}
+
+func (a *FilterArgs) Strings() (result []string) {
+	if a.pos > len(a.args) {
+		return nil
+	}
+	hasErr := false
+	for _, arg := range a.args[a.pos:] {
+		a.pos++
+		if s, ok := arg.(string); ok {
+			result = append(result, s)
+		} else {
+			a.err("%v is not a string", arg)
+			hasErr = true
+		}
+	}
+	if hasErr {
+		return nil
+	}
+	return
+}
+
+func (a *FilterArgs) Float64() (_ float64) {
+	if x, ok := a.next(); ok {
+		switch f := x.(type) {
+		case float64:
+			return f
+		case int:
+			return float64(f)
+		}
+		a.err("%v is not a float64", x)
+	}
+	return
+}
+
+func (a *FilterArgs) Int64() (_ int64) {
+	if x, ok := a.next(); ok {
+		switch i := x.(type) {
+		case int64:
+			return i
+		case int:
+			return int64(i)
+		case float64:
+			ii := int64(i)
+			// check if integer
+			if float64(ii) == i {
+				return ii
+			}
+		}
+		a.err("%v is not an int64", x)
+	}
+	return
+}
+
+func (a *FilterArgs) Int() (_ int) {
+	if x, ok := a.next(); ok {
+		switch i := x.(type) {
+		case int:
+			return i
+		case float64:
+			ii := int(i)
+			// check if integer
+			if float64(ii) == i {
+				return ii
+			}
+		}
+		a.err("%v is not an integer", x)
+	}
+	return
+}
+
+func (a *FilterArgs) IntOr(defValue int) int {
+	if a.pos >= len(a.args) {
+		return defValue
+	}
+	return a.Int()
+}
+
+func (a *FilterArgs) Duration() (_ time.Duration) {
+	if x, ok := a.next(); ok {
+		var d time.Duration
+		switch t := x.(type) {
+		case time.Duration:
+			d = t
+		case int:
+			d = time.Duration(time.Duration(t) * time.Millisecond)
+		case float64:
+			// support fractional milliseconds
+			d = time.Duration(t * float64(time.Millisecond))
+		case string:
+			var err error
+			d, err = time.ParseDuration(t)
+			if err != nil {
+				a.err("%v", err)
+				return
+			}
+		default:
+			a.err("%v is not a duration", x)
+			return
+		}
+		if d < 0 {
+			a.err("duration %v is negative", x)
+			return
+		}
+		return d
```

Comment:
The package should probably expose conversion methods to be used directly if needed

Reply:
Fixed